src/main/java/com/coders/repository/ProductRepository.java

```diff
@@ -0,0 +1,45 @@
+package com.coders.repository;
+
+import com.coders.domain.Product;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public final class ProductRepository {
+
+    private static ProductRepository INSTANCE;
+    private Map<Integer, Product> productMap;
+    private AtomicInteger idGenerator;
+
+    private ProductRepository() {
+        productMap = new HashMap<>();
+        idGenerator = new AtomicInteger(1);
+    }
+
+    public static ProductRepository getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new ProductRepository();
+        }
+
+        return INSTANCE;
+    }
+
+    public void addProduct(Product product) {
+        int productId = idGenerator.getAndIncrement();
```

Comment:
Bardzo dobrze !

Brakuje sprawdzenia co jeśli id nie istnieje w mapie.
Musimy to obsłużyć , stworzyć własne wyjątki.

https://www.baeldung.com/java-new-custom-exception